### ISSUE TITLE
Tensorflow distribute strategy with custom layers gives an error when using Adam optimizer

### ISSUE THREAD

**Issue by capilano** [tensorflow/tensorflow#31724]
`<em>Please` make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Y
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): 
- TensorFlow version (use command below):1.14.0
- Python version:3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I am getting an error when I use a model with custom layers with trainable weights. It works fine without the distribute strategy, and also works fine if the custom layer is replaced by  an existing tf.Keras layer. The error appears to be related to Adam optimizer finding an empty var list when it tries to colocate variables. 

**Describe the expected behavior**

**Code to reproduce the issue**
`
      
     

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by srihari-humbarwadi** [comment]
@capilano Can you please attach a minimal code snippet which can reproduce this issue?

**Comment by capilano** [comment]
This was related to my weights not being created properly when using self.add_variable inside a subclassed Keras layer. I was using adam via tensorflow,  which was returning a weird error, but when I switched to the Keras version, I got a no gradients defined error,which I backtracked to self.add_variable not defining weights as I expected.
My weights were initialized something like this,
self.std = tf.sqrt(2.).
self.custom_weight = self.add_weight(...)*self.std
I changed this to 
self.custom_weight = tf.Variable(init_value*self.std) which solved the problem
The model gets created correctly,the weights are created and reported as trainable when I print the summary(both cases), but throws an error later when the training starts, I think because self.add_weight expects a constant and not a Tensor

**Comment by srihari-humbarwadi** [comment]
Can you please paste the complete code that you are running? And also please paste the weird error logs also. Without that it will hard to say what's going on. @capilano

**Comment by capilano** [comment]
@srihari-humbarwadi  Like I mentioned in the post above, the error was due to the weights not being created properly, and thus there were no gradients to back propogate. 
I solved the problem using tf.Variable instead of layer.add_variable , although both are equivalent, the second one apparently does not work in some cases(like the example I posted above)
When I posted the issue, I felt the problem was elsewhere,because there are no errors generated during the creation of the model and the summary does print out correct parameters. 
I think the issue can be closed, but if you need a code snippet, to reproduce the original issue, I can provide one.

**Comment by ravikyram** [comment]
I am closing this issue since the query is been resolved. Please, feel free to reopen if the issue still persists.Thanks!
